WalletWasabi.Fluent.Generators/UIContextGenerator.cs

```diff
@@ -0,0 +1,332 @@
+
+				var sourceText = SourceText.From(code, Encoding.UTF8);
+				context.AddSource(fileName, sourceText);
+
+				var tree = SyntaxFactory.ParseSyntaxTree(sourceText);
+
+				var newConstructor =
+					tree.GetRoot()
+						.DescendantNodes()
+						.OfType<ConstructorDeclarationSyntax>()
+						.FirstOrDefault();
+
+				yield return newConstructor;
+			}
+		}
+	}
+
+	private static void GenerateFluentNavigation(GeneratorExecutionContext context, IEnumerable<ConstructorDeclarationSyntax> ctors)
+	{
+		var compilation = context.Compilation;
+		var namespaces = new List<string>();
+		var methods = new List<string>();
+
+		var newSyntaxTrees =
+			ctors.Select(x => x.SyntaxTree)
+				 .Where(x => !compilation.ContainsSyntaxTree(x))
+				 .ToList();
+
+		compilation = compilation.AddSyntaxTrees(newSyntaxTrees);
+
+		foreach (var ctor in ctors)
+		{
+			var semanticModel = compilation.GetSemanticModel(ctor.SyntaxTree);
+
+			if (ctor.Parent is not ClassDeclarationSyntax cls)
+			{
+				continue;
+			}
+
+			if (!cls.IsRoutableViewModel(semanticModel))
+			{
+				continue;
+			}
+
+			var viewModelTypeInfo =
+				semanticModel.GetDeclaredSymbol(ctor.Parent as ClassDeclarationSyntax);
+
+			var className = cls.Identifier.ValueText;
+
+			var ctorNamespaces =
+					ctor.ParameterList.Parameters
+									  .Where(p => p.Type is not null)
+									  .Select(p => semanticModel.GetTypeInfo(p.Type!))
+									  .Where(t => t.Type is not null)
+									  .SelectMany(t => t.Type.GetNamespaces())
+									  .ToList();
+
+			var uiContextParam =
+				ctor.ParameterList
+					.Parameters
+					.FirstOrDefault(x => x.Type.IsUiContextType(semanticModel));
+
+			var methodParams = ctor.ParameterList;
+
+			if (uiContextParam != null)
+			{
+				methodParams = SyntaxFactory.ParameterList(methodParams.Parameters.Remove(uiContextParam));
+			}
+
+			var navigationMetadata =
+				viewModelTypeInfo.GetAttributes()
+						.FirstOrDefault(x => x.AttributeClass?.ToDisplayString() == NavigationMetaDataGenerator.NavigationMetaDataAttributeDisplayString);
+
+			var defaultNavigationTarget = "DialogScreen";
+
+			if (navigationMetadata != null)
+			{
+				var navigationArgument =
+					navigationMetadata.NamedArguments
+									  .FirstOrDefault(x => x.Key == "NavigationTarget");
+
+				if (navigationArgument.Value.Type is INamedTypeSymbol navigationTargetEnum)
+				{
+					var enumValue =
+						navigationTargetEnum.GetMembers()
+											.OfType<IFieldSymbol>()
+											.FirstOrDefault(x => x.ConstantValue?.Equals(navigationArgument.Value.Value) ?? false);
+
+					if (enumValue != null)
+					{
+						defaultNavigationTarget = enumValue.Name;
+					}
+				}
+			}
+
+			var additionalMethodParams =
+				$"NavigationTarget navigationTarget = NavigationTarget.{defaultNavigationTarget}, NavigationMode navigationMode = NavigationMode.Normal";
+
+			methodParams = methodParams.AddParameters(SyntaxFactory.ParseParameterList(additionalMethodParams).Parameters.ToArray());
```

Comment:
Instead of parsing a tree at runtime, you can use http://roslynquoter.azurewebsites.net/ to create a method chain. The copy paste it into your code.

In VS there's even a built in tool, but I don't use VS, so don't remember the tool name.

Reply:
@YohDeadfall Normally I would be 100% in favor of a more strongly typed approach and against string typing, but in this specific case of creating an AST node at runtime, I would argue that the string-based approach is much preferable, because not only the roslyn code is extremely larger:

![image](https://user-images.githubusercontent.com/98904713/230168567-2bb4c26d-3a46-40d9-b7a3-10a3ef40256e.png)

It is also much harder to follow and visualize what is the final result from all that compared to the string version

Reply:
You fed the expression as is, so the visualizer thought that it's a variable declaration. The correct code should be that:

```
SeparatedList<ParameterSyntax>(
    new SyntaxNodeOrToken[]{
        Parameter(Identifier("navigationTarget"))
        .WithType(IdentifierName("NavigationTarget"))
        .WithDefault(
            EqualsValueClause(
                MemberAccessExpression(
                    SyntaxKind.SimpleMemberAccessExpression,
                    IdentifierName("NavigationTarget"),
                    IdentifierName(defaultNavigationTarget)))),
        Token(SyntaxKind.CommaToken),
        Parameter(Identifier("navigationMode"))
        .WithType(IdentifierName("NavigationMode"))
        .WithDefault(
            EqualsValueClause(
                MemberAccessExpression(
                    SyntaxKind.SimpleMemberAccessExpression,
                    IdentifierName("NavigationMode"),
                    IdentifierName("Normal"))))})

```

It can be twice shorter if parameter syntax creations is moved to a method accepting 3 parameters: name, type, default.

```
SeparatedList<ParameterSyntax>(
    new SyntaxNodeOrToken[] {
        ParameterWithDefault("navigationTarget", "NavigationTarget", defaultNavigationTarget),
        ParameterWithDefault("navigationMode", "NavigationMode", "Normal"),
    });

static ParameterSyntax ParameterWithDefault(string name, string type, string default) =>
    Parameter(Identifier("navigationMode"))
        .WithType(IdentifierName("NavigationMode"))
        .WithDefault(
            EqualsValueClause(
                MemberAccessExpression(
                    SyntaxKind.SimpleMemberAccessExpression,
                    IdentifierName("NavigationMode"),
                    IdentifierName("Normal"))));
```